### PR TITLE
feat: better assist

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

This PR does the following things:
- refactors `AnalyzerOptions` for better maintainability
- refactors the logic of how the analyzer visitor is created
- refactors the Biome assist
- adds a new feature that allows to run rules outside of the configuration

### `AnalyzerOptions` refactor

When we added new options to `AnalyzerOptions` type, that broke the website job because we had to update the code. I decided to make the fields private and provide builder methods to add the relevant information. 

Once this change will be merged on `main`, it will break the website repo, but at least it won't break anymore in the future.

### New visitor logic

I changed how the Analyzer visitor builder is created. It was really weird to have those methods called `with_linter`, `wiith_assist` that accept the same parameters, so now the publish methods are the parameters themselves, and the method `finish()` creates the visitors itself and returns the enabled/disabled rules. It seems better now.

### Biome assist

There was an improper use of "assists" (plural) and actions. I renamed every "assists" instanced to "assist". This is in line with "Linter" -> "Rules" and "Assist" -> "Actions". 

### New feature

Now linting Workspace methods accept a new `enabled_rules` field. This allows to run analyzer rules outside of the configuration, **on top** of a possible configuration. The reason why we need a new field is because `only` and `skip` have different semantics and have their own specific business logic, which is different from new `rules` field.

This opens up a new feature in our LSP, implemented in this PR, where users can specify `source` code actions that could be applied when formatting. For example, in VSCode, a user now can specify this:

```json
"editor.codeActionsOnSave": {
	"source.biome.useSortedKeys": "explicit"
}
```

On save, the editor will also sort the keys of the JSON objects because we already have that rule implemented for the JSON language. Once and if we implement this rule for JavaScript, it will be applied to that language, too. 

As per LSP, "source" actions are those actions that can be safely applied on format. Since we already have a `source` group, and we already support the `source.biome` capability, it's only sufficient to add the name of the rule

The same thing goes for linter rules! However, the way users would want apply linter rules is a bit different:

```json
"editor.codeActionsOnSave": {
	"quickfix.biome.style.useConst": "explicit"
}
```
Or even 

```json
"editor.codeActionsOnSave": {
	"quickfix.biome.style": "explicit"
}
```

Linter rules are marked as a `quickfix.biome`, so we require the group and the name of the rule.

## Test Plan

I added new tests and tested the new feature locally using VSCode.

### CODE REVIEW

File: .changeset/code_actions_via_ideeditor.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+---
+cli: minor
+---
+
+# Code actions via IDE/editor
+
+Biome can now apply code actions of linter rules or assist actions directly from the settings of their IDE/editor.
```
Comment (arendjr) [inline .changeset/code_actions_via_ideeditor.md:7]:
```suggestion
Biome users can now configure code actions from linter rules as well as assist actions directly in the settings of their IDE/editor.
```

File: .changeset/code_actions_via_ideeditor.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+For example, let's consider the lint rule [`noSwitchDeclarations`](https://biomejs.dev/linter/rules/no-switch-declarations/), which has an unsafe fix.
+Nowadays, if you want to use this rule, you're "forced" to enable it via configuration, and if you want to apply its fix when you save the file, you're forced to make its fix safe, which could be a hazard for the project and team you work with:
```
Comment (arendjr) [inline .changeset/code_actions_via_ideeditor.md:10]:
```suggestion
Previously, if you wanted to use this rule, you were "forced" to enable it via configuration, and if you wanted to apply its fix when you saved a file, you were forced to mark the fix as safe:
```

File: .changeset/code_actions_via_ideeditor.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+          "level": "error",
+          "fix": "safe"
+        }
+      }
+    }
+  }
+}
+```
+
+Now, you can benefit from the code action without making the fix safe for the entire project or for the entire team. IDEs and editor that are compatible LSP allow to list a series of "filters" or code actions that can be applied on save. In the case of VSCode, you will need to add the following snippet in the `settings.json`:
```
Comment (arendjr) [inline .changeset/code_actions_via_ideeditor.md:27]:
```suggestion
Now, you can benefit from the code action without making the fix safe for the entire project. IDEs and editors that are LSP compatible allow to list a series of "filters" or code actions that can be applied on save. In the case of VS Code, you will need to add the following snippet in the `settings.json`:
```

File: .changeset/code_actions_via_ideeditor.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+```json
+{
+  "editor.codeActionsOnSave": {
+    "quickfix.biome.correctness.noSwitchDeclarations": "explicit"
+  }
+}
+```
+
+Upon save, the Biome LSP will tell the editor the apply the code action of the rule `noSwitchDeclarations`.
```
Comment (arendjr) [inline .changeset/code_actions_via_ideeditor.md:37]:
```suggestion
Upon save, Biome will inform the editor the apply the code action of the rule `noSwitchDeclarations`.
```

File: crates/biome_analyze/src/options.rs [hunk 5]
```diff
@@ -55,34 +55,76 @@ impl AnalyzerRules {
...
+
+    pub fn with_preferred_quote(mut self, preferred_quote: PreferredQuote) -> Self {
+        self.preferred_quote = preferred_quote;
+        self
+    }
+
+    /// Returns the rules that are configured for this analyzer
+    pub fn rules(&self) -> &AnalyzerRules {
+        &self.rules
+    }
```
Comment (arendjr) [inline crates/biome_analyze/src/options.rs:96]:
```suggestion
    /// Returns the rules that are configured for this analyzer
    pub fn rules(&self) -> &AnalyzerRules {
        &self.rules
    }
```

I think it's both simpler and more flexible to let the `rules` field remain public and then you wouldn't need this getter.

Rust ownership rules already prevent you from messing with instances you may not mutate, and if there are no additional business rules to enforce (there don't seem to be any here), then there's no real reason to do it like this.

File: crates/biome_analyze/src/options.rs [hunk 6]
```diff
@@ -117,6 +159,14 @@ impl AnalyzerOptions {
...
         &self.configuration.preferred_quote
     }
+
+    pub fn file_path(&self) -> &Path {
+        self.file_path.as_ref()
+    }
+
+    pub fn suppression_reason(&self) -> Option<&str> {
+        self.suppression_reason.as_deref()
+    }
```
Comment (arendjr) [inline crates/biome_analyze/src/options.rs:169]: Same for these getters.

File: crates/biome_js_analyze/src/assist.rs [hunk 7]
```diff
@@ -0,0 +1,4 @@
+//! Generated file, do not edit by hand, see `xtask/codegen`
```
Comment (arendjr) [inline crates/biome_js_analyze/src/assist.rs:1]: I think `.gitattributes` should still be updated?

File: crates/biome_json_analyze/src/assist.rs [hunk 8]
```diff
@@ -0,0 +1,4 @@
+//! Generated file, do not edit by hand, see `xtask/codegen`
```
Comment (arendjr) [inline crates/biome_json_analyze/src/assist.rs:1]: For this one too.

**Review by arendjr — APPROVED** [review]
Nice! Mainly some copy suggestions.

**Review by Conaclos — APPROVED** [review]

File: crates/biome_lsp/tests/server.rs [hunk 9]
```diff
@@ -1916,6 +1916,7 @@ async fn does_not_pull_action_for_disabled_rule_in_override_issue_2782() -> Resu
         .await?
         .context("codeAction returned None")?;
 
+    dbg!(&res);
```
Comment (dyc3) [inline crates/biome_lsp/tests/server.rs:1919]:
```suggestion
```

Comment (ematipico) [reply crates/biome_lsp/tests/server.rs:1919]: Because it's allowed inside tests :)

**Pull request merged**